cmd/utils.go

```diff
@@ -686,18 +687,45 @@ func ceilFrac(numerator, denominator int64) (ceil int64) {
 	return
 }
 
-// Returns context with ReqInfo details set in the context.
-func newContext(r *http.Request, w http.ResponseWriter, api string) context.Context {
-	vars := mux.Vars(r)
-	bucket := vars["bucket"]
-	object, err := url.PathUnescape(vars["object"])
+// unescapeGeneric is similar to url.PathUnescape or url.QueryUnescape
+// depending on input, additionally also handles situations such as
+// `//` are normalized as `/`, also removes any `/` prefix before
+// returning.
+func unescapeGeneric(p string, escapeFn func(string) (string, error)) (string, error) {
+	ep, err := escapeFn(p)
 	if err != nil {
-		object = vars["object"]
+		ep = p
 	}
-	prefix, err := url.QueryUnescape(vars["prefix"])
+	if len(ep) > 0 && ep[0] == '/' {
+		ep = path.Clean(ep)
```

Comment:
This will remove the trailing slash if any, I wonder how unit tests didn't catch object-dir testing

Reply:
Mainly because we are only doing this if `ep[0]` has a `/` in the beginning - to preserve that we need to add retainSlash() code here.